vm/src/stdlib/pystruct.rs

```diff
@@ -252,171 +279,317 @@ fn pack_bool(vm: &VirtualMachine, arg: &PyObjectRef, data: &mut dyn Write) -> Py
+    let mut buf: Vec<u8> = Vec::new();
+    handle.read_to_end(&mut buf).map_err(|_| {
+        new_struct_error(vm, format!("unpack requires a buffer of {} bytes", length,))
+    })?;
+    Ok(vm.ctx.new_bytes(buf))
 }
 
 fn struct_unpack(fmt: PyStringRef, buffer: PyBytesRef, vm: &VirtualMachine) -> PyResult<PyTuple> {
     let fmt_str = fmt.as_str();
-    let format_spec = FormatSpec::parse(fmt_str).map_err(|e| vm.new_value_error(e))?;
+    let format_spec = FormatSpec::parse(fmt_str).map_err(|e| new_struct_error(vm, e))?;
     format_spec.unpack(buffer.get_value(), vm)
 }
 
-fn unpack_code<Endianness>(vm: &VirtualMachine, code: &FormatCode, rdr: &mut dyn Read) -> PyResult
+fn unpack_code<Endianness>(
+    vm: &VirtualMachine,
+    code: &FormatCode,
+    rdr: &mut dyn Read,
+    items: &mut Vec<PyObjectRef>,
+) -> PyResult<()>
 where
     Endianness: byteorder::ByteOrder,
 {
-    match code.code {
-        'b' => unpack_i8(vm, rdr),
-        'B' => unpack_u8(vm, rdr),
-        '?' => unpack_bool(vm, rdr),
-        'h' => unpack_i16::<Endianness>(vm, rdr),
-        'H' => unpack_u16::<Endianness>(vm, rdr),
-        'i' | 'l' => unpack_i32::<Endianness>(vm, rdr),
-        'I' | 'L' => unpack_u32::<Endianness>(vm, rdr),
-        'q' => unpack_i64::<Endianness>(vm, rdr),
-        'Q' => unpack_u64::<Endianness>(vm, rdr),
-        'f' => unpack_f32::<Endianness>(vm, rdr),
-        'd' => unpack_f64::<Endianness>(vm, rdr),
+    let unpack = match code.code {
+        'b' => unpack_i8,
+        'B' => unpack_u8,
+        'c' => unpack_char,
+        '?' => unpack_bool,
+        'h' => unpack_i16::<Endianness>,
+        'H' => unpack_u16::<Endianness>,
+        'i' | 'l' => unpack_i32::<Endianness>,
+        'I' | 'L' => unpack_u32::<Endianness>,
+        'q' => unpack_i64::<Endianness>,
+        'Q' => unpack_u64::<Endianness>,
+        'n' => unpack_isize::<Endianness>,
+        'N' => unpack_usize::<Endianness>,
+        'P' => unpack_usize::<Endianness>, // FIXME: native-only
+        'f' => unpack_f32::<Endianness>,
+        'd' => unpack_f64::<Endianness>,
+        'x' => {
+            unpack_empty(vm, rdr, code.repeat);
+            return Ok(());
+        }
+        's' | 'p' => {
+            items.push(unpack_string(vm, rdr, code.repeat)?);
+            return Ok(());
+        }
         c => {
             panic!("Unsupported format code {:?}", c);
         }
+    };
+    for _ in 0..code.repeat {
+        items.push(unpack(vm, rdr)?);
     }
+    Ok(())
 }
 
-fn struct_calcsize(fmt: PyStringRef, vm: &VirtualMachine) -> PyResult<usize> {
-    let fmt_str = fmt.as_str();
-    let format_spec = FormatSpec::parse(fmt_str).map_err(|e| vm.new_value_error(e))?;
+fn struct_calcsize(fmt: Either<PyStringRef, PyBytesRef>, vm: &VirtualMachine) -> PyResult<usize> {
+    let parsed = match fmt {
+        Either::A(string) => FormatSpec::parse(string.as_str()),
+        Either::B(bytes) => FormatSpec::parse(std::str::from_utf8(&bytes).unwrap()),
```

Comment:
It looks like CPython primarily parses on bytes, and encodes `str`s to bytes using the ascii codec:
```python
>>> struct.calcsize(">BȔ")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'ascii' codec can't encode character '\u0214' in position 2: ordinal not in range(128)
```

Reply:
I added a comment about this, but didn't fix it at this time.